### PR TITLE
Fixes #7462 - new UI for network interfaces

### PR DESCRIPTION
**Pull request by tstrachota**
Editing network interfaces moved into a modal window. Some screenshots of the new UI are at:
https://tstrachota.fedorapeople.org/screenshots/nics_ui/

Bits of host form related to NICs are initially hidden and copied into a modal window on edit button click. This allows reusing validation error notices rendered by rails.

### CODE REVIEW

**Comment by tstrachota** [comment]
@ares @GregSutcliffe

File: app/views/nic/_hidden_layout.html.erb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+<div style="display: none;" id="interfaceHidden<%= local_assigns[:id] %>" interface-id="<%= local_assigns[:id] %>">
```
Comment (ohadlevy) [inline app/views/nic/_hidden_layout.html.erb]: nitpick: use class=hide

File: app/assets/javascripts/host_edit_interfaces.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  return interface_hidden;
+}
+
+function update_interface_row(row, modal_window) {
+  row.find('.type').html(modal_window.find('.interface_type option:selected').text());
+  row.find('.identifier').html(modal_window.find('.interface_identifier').val());
+  row.find('.mac').html(modal_window.find('.interface_mac').val());
+}
+
+
```
Comment (ares) [inline app/assets/javascripts/host_edit_interfaces.js]: nitpick: few extra lines

File: app/assets/javascripts/host_edit_interfaces.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    form = $('#interfaces #interfaceHidden'+interface_id).clone();
+
+  show_interface_modal(form);
+}
+
+function show_interface_modal(modal_content) {
+  var modal_window = $('#interfaceModal');
+  var interface_id = modal_content.attr('interface-id');
+
+  modal_window.attr('current-id', interface_id);
```
Comment (ares) [inline app/assets/javascripts/host_edit_interfaces.js]: I think using data attributes would be better for current-id

**Comment by ares** [comment]
Could you please add some space between green button "Add Interface" and horizontal line below?

**Comment by ares** [comment]
I'm not sure whether this issue is introduced by this PR but when I try to add new interface to existing host and fill in only identifier, it allows me to save the host, I see error but the interface is not created. Same applies to new host.

**Comment by tstrachota** [comment]
Updated:
- spacing around `+ Add interface` button
- link up/down status icons
- background of invalid interface's row is now light red
- red title and message when there's a validation error in one of the interfaces
- changed class of interface edit button

Screenshots updated too.

**Comment by tstrachota** [comment]
@ares the issue you came across with is an error somewhere in validations (it's in development branch too). NICs with blank mac addresses are just removed.

**Comment by ares** [comment]
@tstrachota definitely much nicer! Last thing I doubt is the warning icon in edit button (sorry for potential bad hint, I didn't image varying width of button), let's see what others think. Don't worry about the validation issue then, validations may change later and it's not part of this PR.

**Comment by ares** [comment]
@tstrachota you made rubocop angry http://ci.theforeman.org/job/test_develop_pr_rubocop/1054/checkstyleResult/source.0/#332 :-)

**Comment by tstrachota** [comment]
@ares I'm not fond of the varying size too. We can remove the icon. No strong opinion from my side here.

**Comment by ohadlevy** [comment]
IMHO the background color on error is too much. Can we just mark the tr border red (similar to other errors in input fields)?

I think the icon next to edit on error is not needed.maybe change the btn class instead of an icon for f its needed?

**Comment by tstrachota** [comment]
@ohadlevy https://tstrachota.fedorapeople.org/screenshots/nics_ui/005_nics_validation_error_v3.png
Better?

**Comment by ohadlevy** [comment]
yes, I wonder if we should have the warning (above the table) wrapped in a
warning box?

**Comment by tstrachota** [comment]
@ohadlevy updated https://tstrachota.fedorapeople.org/screenshots/nics_ui/006_nics_validation_error_v4.png

**Comment by ohadlevy** [comment]
looks better!, does the buttons look better when you pull-right them?
+1 from my side.

**Comment by tstrachota** [comment]
Buttons on right seem too far from the rest of the content to me.

File: app/views/hosts/_interfaces.html.erb [hunk 4]
```diff
@@ -0,0 +1,86 @@
+
+<% @interfaces_invalid = @host.interfaces.any?{ |i| !i.valid? } %>
```
Comment (ares) [inline app/views/hosts/_interfaces.html.erb]: nitpick: could you please use interfaces_invalid without @ so we don't risk any conflict in future?

File: app/views/nic/new.js.erb [hunk 5]
```diff
@@ -1,8 +1,8 @@
 <% if @interface.present? %>
   <%= fields_for :host, Host::Base.new do |host| %>
     <%= host.fields_for :interfaces, @interface, :child_index => @key do |interfaces| %>
-      $("#host_interfaces_attributes_<%= @key %>_type").closest('fieldset').html(
-        '<%=j render :partial => @interface, :layout => 'nic/interface_layout', :locals => {:f => interfaces} %>'
```
Comment (ares) [inline app/views/nic/new.js.erb:5]: it seems that we don't use nic/interface_layout partial any more so maybe we should remove it?

**Comment by ares** [comment]
@tstrachota I'm happy with the code, could you please squash before I'll reapply and test it?

**Comment by tstrachota** [comment]
@ares done

File: app/views/hosts/_interfaces.html.erb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      <th class="ellipsis"><%= _('Identifier') %></th>
+      <th class="hidden-xs"><%= _('Type') %></th>
+      <th class="hidden-xs"><%= _("MAC address") %></th>
+      <th></th>
+    </tr>
+    <% @host.interfaces.each do |interface| %>
+      <tr class="<%= 'has-error' unless interface.valid? %>" id="interface<%= interface.object_id %>" data-interface-id="<%= interface.object_id %>">
+        <td class="status hidden-xs" align="center"><%= link_status(interface) %></td>
+        <td class="identifier ellipsis"><%= interface.identifier %></td>
+        <td class="type hidden-xs"><%= interface.class.humanized_name %></td>
```
Comment (ares) [inline app/views/hosts/_interfaces.html.erb]: missing N_()

**Comment by ares** [comment]
Merged as https://github.com/theforeman/foreman/commit/3154d77e023e84a2515c6a9bdcaf974fcdbb4493, thanks @tstrachota!

**Pull request closed without merging**